.github/workflows/samples-tools/cmd/run-all/main.go

```diff
@@ -0,0 +1,88 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"os/exec"
+	"samples-tools/pkg/utils"
+	"strings"
+)
+
+func main() {
+	configFile := flag.String("config", "", "path to the config file")
+	parallel := flag.Bool("parallel", false, "set to true to run tests in parallel")
+	flag.Parse()
+
+	if *configFile == "" {
+		fmt.Fprintf(os.Stderr, "config file is required, please pass -config=path/to/config.jsonc\n")
+		os.Exit(1)
+	}
+
+	cmd := flag.Arg(0)
+	if cmd == "" {
+		fmt.Fprintf(os.Stderr, "command is required, use positional arguments with %%s for the package placeholder\n")
+		os.Exit(1)
+	}
+
+	args := flag.Args()[1:]
+
+	config, err := utils.LoadConfig(*configFile)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error loading the config file, make sure it exists and it's valid: %v\n", err)
+		os.Exit(1)
+	}
+
+	packages, err := utils.FindAllPackages(".", config)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error finding packages: %v\n", err)
+	}
+
+	failed := runAll(packages, cmd, args, *parallel)
+	fmt.Printf("Total tests: %v\nFailed tests: %v\n", len(packages), failed)
+	if failed > 0 {
+		os.Exit(1)
+	}
+}
+
+func runAll(packages []string, cmd string, args []string, parallel bool) int {
+	failures := utils.Map(parallel, packages, func(pkg string) string {
+		return runOne(pkg, cmd, args)
+	})
+
+	failed := 0
+	for _, failure := range failures {
+		if failure != "" {
+			failed++
+		}
+	}
+	return failed
+}
+
+func runOne(pkg string, cmd string, args []string) string {
+	// TODO(dcavazos): measure time spent on each test and print it along the results
+	cmdArgs := make([]string, len(args))
+	for i, arg := range args {
+		if strings.Contains(arg, "%s") {
+			cmdArgs[i] = fmt.Sprintf(arg, pkg)
+		} else {
+			cmdArgs[i] = arg
+		}
+	}
+	output, err := exec.Command(cmd, cmdArgs...).CombinedOutput()
```

Comment:
comment: So, I'm a bit torn on this point.

As I learned Go, the model was to prefer creating goroutines for executing functions and use channels to communicate. I don't think this is wrong, per se, but a little odd (to me at any rate).

Reply:
We're running `make test dir=${package}` here. Is there a way in Go to run a shell command without `exec`? We would still be using goroutines to parallelize the run, but each test is done through the existing Makefile entrypoint.